### PR TITLE
Misc cleanups to verbs send functions.

### PR DESCRIPTION
**Pull request by pmmccorm**
Don't pass around length arguments and then just ignore
them. Fill out WR in upper layer fi_ibv_send* functions
and pass down to actual verbs send.

Don't drop other send flags when given FI_FENCE.

Signed-off-by: pmmccorm <patrick.m.mccormick@intel.com>

### CODE REVIEW

**Comment by pmmccorm** [comment]
Re-spun and rebased to try to cajole a Travis build.

**Comment by pmmccorm** [comment]
Are there any other changes needed for this PR? I brought back the macros.

**Comment by pmmccorm** [comment]
Pushed the SOCK_STREAM_EP as separate commit. @shefty @a-ilango can you take a look and tell me if this approach looks sensible?

**Comment by shefty** [comment]
See build errors in travis

File: prov/verbs/src/verbs_msg_ep.c [hunk 1]
```diff
@@ -238,7 +215,13 @@ static int fi_ibv_msg_ep_enable(struct fid_ep *ep)
 	}
 
 	attr.qp_type = IBV_QPT_RC;
-	attr.sq_sig_all = 0;
+
+	/* assuming FI_EP_SOCK_STREAM */
```
Comment (shefty) [inline prov/verbs/src/verbs_msg_ep.c]: Add an assert() or validation check

File: prov/verbs/src/verbs_msg.c [hunk 2]
```diff
@@ -212,6 +345,19 @@ struct fi_ops_msg fi_ibv_msg_ep_msg_ops = {
...
+	.size = sizeof(struct fi_ops_msg),
+	.recv = fi_ibv_sstream_ep_recv,
+	.recvv = fi_ibv_sstream_ep_recvv,
+	.recvmsg = fi_ibv_sstream_ep_recvmsg,
+	.send = fi_ibv_sstream_ep_send,
+	.sendv = fi_ibv_sstream_ep_sendv,
+	.sendmsg = fi_ibv_sstream_ep_sendmsg,
+	.inject = fi_ibv_msg_ep_inject,
+	.senddata = fi_no_msg_senddata,
+	.injectdata = fi_no_msg_injectdata,
```
Comment (shefty) [inline prov/verbs/src/verbs_msg.c]: injectdata and senddata are not supported by stream eps, since there's no peer CQ to write the data to.

File: prov/verbs/src/verbs_msg.c [hunk 3]
```diff
@@ -127,6 +211,21 @@ fi_ibv_msg_ep_sendmsg(struct fid_ep *ep_fid, const struct fi_msg *msg, uint64_t
...
+fi_ibv_sstream_ep_sendmsg(struct fid_ep *ep_fid, const struct fi_msg *msg, uint64_t flags)
+{
+	ssize_t ret;
+
+	ret = fi_ibv_msg_ep_sendmsg(ep_fid, msg, flags);
+	if (ret)
+		return ret;
+
+	if ((ret = fi_ibv_get_send_comp(ep_fid)) == 0)
+		return fi_ibv_msg_total(msg);
```
Comment (shefty) [inline prov/verbs/src/verbs_msg.c]: This semantic isn't going to work.  Verbs requires that the send and receive buffers be registered before being used.  There's no requirement for registration for the user of a stream EP.  The data will need to be copied into a pre-registered buffer prior to sending.

Additionally, blocking every send operation until a completion has been posted will result in horrible performance.  Completions are written by verbs hardware until the data has been acked by the target.  That's a significant delay.  We want to emulate nonblocking sockets, since that's how most applications use them.

File: prov/verbs/src/verbs_msg.c [hunk 4]
```diff
@@ -30,52 +30,104 @@
...
+fi_ibv_sstream_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, uint64_t flags)
+{
+	ssize_t ret;
+
+	ret = fi_ibv_msg_ep_recvmsg(ep, msg, flags);
+	if (ret)
+		return ret;
+
+	if ((ret = fi_ibv_get_recv_comp(ep)) == 0)
+		return fi_ibv_msg_total(msg);
```
Comment (shefty) [inline prov/verbs/src/verbs_msg.c]: See comment below regarding memory registration and completion issues.

Additionally, this won't support nonblocking socket semantics.  The sender cannot transfer data until the receiver has actually posted the receive buffer.  Otherwise flow control will back the sender off until a buffer has been posted. The receive buffer must be made available to the sender prior to the application calling receive.  Recv() is used to retrieve locally buffered data, not provide network buffers.

**Comment by shefty** [comment]
The approach won't work.  Please look at the rsocket implementation regarding how data needs to be sent and received.  Trying to use the application's buffers directly with the network won't be viable.

File: prov/verbs/src/verbs_msg.c (hunk 2 above)
Comment (pmmccorm) [reply prov/verbs/src/verbs_msg.c]: This patch reflects that right?

File: prov/verbs/src/verbs_msg.c (hunk 3 above)
Comment (pmmccorm) [reply prov/verbs/src/verbs_msg.c]: I see, so what would a non-blocking usage of SOCK_STREAM_EP look like?

File: prov/verbs/src/fi_verbs.c [hunk 5]
```diff
@@ -357,12 +363,14 @@ ssize_t fi_ibv_send_iov_flags(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr,
...
 		fi_ibv_set_sge_iov(wr->sg_list, iov, count, desc, len);
 
-	wr->send_flags = VERBS_INJECT_FLAGS(ep, len, flags) | VERBS_COMP_FLAGS(ep, flags);
+	wr->num_sge = count;
+	wr->send_flags |= VERBS_INJECT_FLAGS(ep, len, flags);
+	wr->send_flags |= VERBS_COMP_FLAGS(ep, flags);
 
 	if (flags & FI_FENCE)
-		wr->send_flags = IBV_SEND_FENCE;
+		wr->send_flags |= IBV_SEND_FENCE;
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: This bug fix still needs to go into a separate patch.

Comment (pmmccorm) [reply prov/verbs/src/fi_verbs.c]: Done in new version.

File: prov/verbs/src/verbs_msg_ep.c (hunk 1 above)
Comment (a-ilango) [reply prov/verbs/src/verbs_msg_ep.c]: ^ +1

The EP type should be checked and further checks can be done based on that. If needed, the checks can be put in a separate function. We should not remove the existing checks for msg EP type.

**Comment by pmmccorm** [comment]
New version: remove SOCK_STREAM stuff (will review rsocket implementation and make separate PR), put FI_INJECT and cleanups in separate patches.

**Comment by a-ilango** [comment]
I guess my comments on SOCKSTREAM EP got added after the PR was updated.

**Comment by pmmccorm** [comment]
I will make the suggested changes and re-post.

File: prov/verbs/src/verbs_domain.c [hunk 6]
```diff
@@ -255,16 +255,12 @@ static int fi_ibv_open_device_by_name(struct fi_ibv_domain *domain, const char *
...
-		if (domain->rdm) {
-			ret = strncmp(name, ibv_get_device_name(dev_list[i]->device),
-				      strlen(name) - strlen(verbs_rdm_domain.suffix));
-
-		} else {
-			ret = strcmp(name, ibv_get_device_name(dev_list[i]->device));
-		}
-
-		if (!ret)
+		const char *rdma_name = ibv_get_device_name(dev_list[i]->device);
```
Comment (a-ilango) [inline prov/verbs/src/verbs_domain.c]: nit: please move variable declarations to the top.

**Comment by pmmccorm** [comment]
Rebased on top of current master, added another trivial fix in a separate commit.
